### PR TITLE
[#71] [Backend] As a user, I can see the home header

### PR DESCRIPTION
**Pull request by markgravity**
https://github.com/markgravity/kmm-ic/issues/71

## What happened 👀

- Refactor `AuthRepository` to use `TokenService`
- Support Authentication in `ApiService`
- Setup test environment
- Write Unit Test
- Add `UserService`, `UserRepository` & `GetCurrentUserUseCase`

## Insight 📝

n/a

## Proof Of Work 📹

CleanShot 2023-04-21 at 13 49 14

### CODE REVIEW

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/user/UserService.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import co.nimblehq.mark.kmmic.data.service.api.extension.path
+import co.nimblehq.mark.kmmic.data.service.user.model.UserResponse
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.http.HttpMethod
+import kotlinx.coroutines.flow.Flow
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+internal interface UserService {
+    fun me(): Flow<UserResponse>
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/user/UserService.kt]: Maybe improve the naming?
```suggestion
    fun getProfile(): Flow<UserResponse>
```

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/model/User.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package co.nimblehq.mark.kmmic.domain.model
+
+import co.nimblehq.mark.kmmic.data.service.user.model.UserResponse
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class User constructor(
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/model/User.kt]: We could omit contructor
```suggestion
data class User(
```

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetCurrentUserUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import org.koin.core.component.inject
+
+interface GetCurrentUserUseCase {
+    fun call(): Flow<User>
+}
+
+class GetCurrentUserUseCaseImpl : GetCurrentUserUseCase, KoinComponent {
+    private val userRepository: UserRepository by inject()
+
+    override fun call(): Flow<User> {
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetCurrentUserUseCase.kt]: Please use `invoke`

**Review by suho — CHANGES_REQUESTED** [review]

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetCurrentUserUseCase.kt (hunk 3 above)
Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetCurrentUserUseCase.kt]: @suho updated at [3225037](https://github.com/markgravity/kmm-ic/pull/72/commits/3225037d11388c90ad505fdf1de85c0ac1d7d6f2)

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/model/User.kt (hunk 2 above)
Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/model/User.kt]: @suho updated at [3225037](https://github.com/markgravity/kmm-ic/pull/72/commits/3225037d11388c90ad505fdf1de85c0ac1d7d6f2)

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/user/UserService.kt (hunk 1 above)
Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/user/UserService.kt]: @suho updated at [3225037](https://github.com/markgravity/kmm-ic/pull/72/commits/3225037d11388c90ad505fdf1de85c0ac1d7d6f2)

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetCurrentUserUseCase.kt (hunk 3 above)
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetCurrentUserUseCase.kt]: @markgravity Please pay attention when updating the code, `invoke` should be use with operator in kotlin.

Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetCurrentUserUseCase.kt]: @suho Please ignore it for now, I will create a chore PR to rename the `invoke` and `callAsFunction` at once 🙏

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/di/DI.kt [hunk 4]
```diff
@@ -1,32 +1,48 @@
...
 
-    val useCaseModule = module {
+    val domainModule = module {
         factory<LoginUseCase> { LoginUseCaseImpl() }
+        factory<GetCurrentUserUseCase> { GetCurrentUserUseCaseImpl() }
     }
 
     startKoin {
-        modules(listOf(commonModule, serviceModule, useCaseModule))
+        modules(listOf(commonModule, dataModule, domainModule))
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/di/DI.kt:46]: Optional: we can concatenate the lists like this in Kotlin:
```suggestion
        modules(commonModule + dataModule + domainModule)
```

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Test
+    fun `when me is called - it returns the correctly user`() = runTest {
+        given(mockUserService)
+            .function(mockUserService::getProfile)
+            .whenInvoked()
+            .thenReturn(
+                flow {
+                    emit(mockUser)
+                }
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt:61]: This is a shorter way:
```suggestion
                flowOf(mockUser)
```

**Review by suho — APPROVED** [review]

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            )
+
+        repository.getProfile().collect {
+            it shouldBe User(mockUser)
+        }
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt:66]: Here's an easy way to assert flow results:
```suggestion
        repository.getProfile().first() shouldBe User(mockUser)
```
We can do the same in other classes too.

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import org.koin.core.context.startKoin
+import org.koin.dsl.module
+import kotlin.test.Test
+import kotlinx.coroutines.test.runTest
+import org.koin.core.context.stopKoin
+import kotlin.test.AfterTest
+
+@ExperimentalCoroutinesApi
+class UserRepositoryTest {
+    @Mock
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt:24]: It'd be nice to have a line break after the class declaration following [this](https://nimblehq.co/compass/development/code-conventions/kotlin/#convention-4) convention. This should apply to all the other classes such as repositories, services, etc.

**Review by phongvhd93 — APPROVED** [review]

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/di/DI.kt (hunk 4 above)
Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/di/DI.kt:46]: @AVI5HEK updated at [e796eb2](https://github.com/markgravity/kmm-ic/pull/72/commits/e796eb2998575c9a9fa82a0f9041c6b3db961ce0)

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt (hunk 5 above)
Comment (markgravity) [reply shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt:61]: @AVI5HEK updated at [e796eb2](https://github.com/markgravity/kmm-ic/pull/72/commits/e796eb2998575c9a9fa82a0f9041c6b3db961ce0)

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt (hunk 6 above)
Comment (markgravity) [reply shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt:66]: @AVI5HEK updated at [e796eb2](https://github.com/markgravity/kmm-ic/pull/72/commits/e796eb2998575c9a9fa82a0f9041c6b3db961ce0)

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt (hunk 7 above)
Comment (markgravity) [reply shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/UserRepositoryTest.kt:24]: @AVI5HEK updated at [e796eb2](https://github.com/markgravity/kmm-ic/pull/72/commits/e796eb2998575c9a9fa82a0f9041c6b3db961ce0)

**Review by AVI5HEK — APPROVED** [review]
LGTM

**Pull request merged**